applications/nrf5340_audio/src/audio_module/amod_api.c

```diff
@@ -0,0 +1,1136 @@
+							&object_release_cb);
+					if (ret) {
+						LOG_DBG("Failed to send to module %s from %s",
+							hdl_to->name, hdl->name);
+					} else {
+						k_sem_give(&hdl->sem);
+					}
+				}
+
+				k_mutex_unlock(&hdl->dest_mutex);
+			} else {
+				/* Send input audio object */
+				ret = data_fifo_block_lock(&hdl->out_msg, (void **)&out_msg,
+							   sizeof(struct _amod_in_message));
+				if (ret) {
+					clean_up(hdl, &in_msg, &out_msg, &data);
+
+					LOG_DBG("Module %s failed to return output buffer, ret %d",
+						hdl->name, ret);
+					continue;
+				}
+				LOG_DBG("Data sent from module %s", hdl->name);
+			}
+
+		} else {
+			LOG_DBG("No process function for module %s, discarding input", hdl->name);
+		}
+	}
+
+	return 0;
+}
+
+/**
+ * @brief The thread that processes inputs and outputs them out of the audio system.
+ *
+ * @param handle  The handle for this modules instance
+ *
+ * @return 0 if successful, error value
+ */
+static int module_thread_output(struct amod_handle *handle)
+{
+	struct _amod_handle *hdl = (struct _amod_handle *)handle;
+	int ret;
+	struct _amod_in_message *in_msg;
+	struct _amod_out_message *out_msg = NULL;
+	char *data = NULL;
+	size_t size;
+
+	if (handle == NULL) {
+		LOG_DBG("Error in the module task as handle NULL");
+		return -EINVAL;
+	}
+
+	/* Execute thread */
+	for (;;) {
+		in_msg = NULL;
+
+		ret = data_fifo_pointer_last_filled_get(&hdl->in_msg, (void **)&in_msg, &size,
+							K_FOREVER);
+		if (ret) {
+			LOG_DBG("No new data for module %s, ret %d", hdl->name, ret);
+			continue;
+		}
+
+		if (in_msg->object == NULL) {
+			clean_up(hdl, &in_msg, &out_msg, &data);
+
+			LOG_DBG("Error in input data for module %s", hdl->name);
+			continue;
+		}
+
+		if (hdl->functions->data_process != NULL) {
+			/* Process the input audio object and output from the audio system */
+			ret = hdl->functions->data_process(handle, in_msg->object, NULL);
+			if (ret) {
+				clean_up(hdl, &in_msg, &out_msg, &data);
+
+				LOG_DBG("Data process error in module %s, ret %d", hdl->name, ret);
+				continue;
+			}
+
+			if (in_msg->response_cb != NULL) {
+				in_msg->response_cb(in_msg->tx_handle, in_msg->object);
+			}
+
+		} else {
+			LOG_DBG("No process function for module %s, discarding input", hdl->name);
+		}
+
+		data_fifo_block_free(&hdl->in_msg, (void **)&in_msg);
+	}
+
+	return 0;
+}
+
+/**
+ * @brief The thread that processes inputs and returns data form the module
+ *
+ * @param handle  The handle for this modules instance
+ *
+ * @return 0 if successful, error value
+ */
+static int module_thread_processor(struct amod_handle *handle)
+{
+	struct _amod_handle *hdl = (struct _amod_handle *)handle;
+	struct _amod_handle *hdl_to;
+	int ret;
+	struct _amod_in_message *in_msg;
+	struct _amod_out_message *out_msg;
+	char *data = NULL;
+	size_t size;
+
+	if (handle == NULL) {
+		LOG_DBG("Error in the module task as handle NULL");
+		return -EINVAL;
+	}
+
+	/* Execute thread */
+	for (;;) {
```

Comment:
Do I get it correctly? : To me it seems that this general thread will execute the processing of all modules?

Reply:
Well the input and output threads are simpler as they either do not pass any data to or get data from another module (e.g. an output peripheral).